docs/man/man3/AELogger.3
.TH "AELogger" 3 "Mon Mar 18 2024 18:44:24" "Version v0.0.8.5a" "ArtyK's Console Engine" \" -*- nroff -*-
.ad l
.nh
.SH NAME
AELogger \- The ArtyK's Engine's Logger module -- it manages the writing to the log files\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fR#include <AELogger\&.hpp>\fP
.PP
Inherits \fB__AEModuleBase< AELogger >\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBAELogger\fP (const std::string fname, const bool clearLog=false, const \fBullint\fP queuesize=\fBAELOG_DEFAULT_QUEUE_SIZE\fP)"
.br
.RI "Class constructor -- it opens the file with the given name and starts logging to it\&. "
.ti -1c
.RI "\fBAELogger\fP (const std::string logpath, const std::string fname, const bool clearLog=false, const \fBullint\fP queuesize=\fBAELOG_DEFAULT_QUEUE_SIZE\fP)"
.br
.RI "Class constructor -- it opens the file with the given path and name and start logging to it\&. "
.ti -1c
.RI "\fBAELogger\fP (void)"
.br
.RI "Default class constructor -- doesn't open the file, but setups the instance for one\&. "
.ti -1c
.RI "\fB~AELogger\fP (void)"
.br
.RI "Class destructor\&. "
.ti -1c
.RI "\fBAELogger\fP (const \fBAELogger\fP &)=delete"
.br
.RI "Deleted (for now) copy-constructor\&. "
.ti -1c
.RI "\fBAELogger\fP & \fBoperator=\fP (const \fBAELogger\fP &)=delete"
.br
.RI "Deleted (for now) copy-assignment\&. "
.ti -1c
.RI "\fBcint\fP \fBstartWriter\fP (void)"
.br
.RI "Starts the log-writing thread\&. "
.ti -1c
.RI "\fBcint\fP \fBstopWriter\fP (void)"
.br
.RI "Stops the log-writing thread (after flushing the log queue)\&. "
.ti -1c
.RI "\fBcint\fP \fBopenLog\fP (const std::string fpath, const std::string fname, const bool clearLog=false)"
.br
.RI "Open the file to start logging\&. "
.ti -1c
.RI "\fBcint\fP \fBopenLog\fP (const std::string fname, const bool clearLog=false)"
.br
.RI "Open the file to start logging\&. "
.ti -1c
.RI "\fBcint\fP \fBopenLog\fP (const std::string_view fpath, const std::string_view fname, const bool clearLog=false)"
.br
.RI "Open the file to start logging\&. "
.ti -1c
.RI "\fBcint\fP \fBcloseLog\fP (void)"
.br
.RI "Close the file, if it was opened\&. "
.ti -1c
.RI "\fBcint\fP \fBwriteToLog\fP (const std::string_view logmessg, const \fBcint\fP logtype=\fBAELOG_TYPE_INFO\fP, const std::string_view logmodule=\fBAELOG_DEFAULT_MODULE_NAME\fP, const std::time_t logTime=0)"
.br
.RI "Request a log entry to be written to the opened log file\&. "
.ti -1c
.RI "\fBcint\fP \fBwriteToLog\fP (const \fBAELogEntry\fP &entry, const bool useCurrentTime=true)"
.br
.RI "Request a log entry to be written to the opened log file, using data from a log entry type\&. "
.ti -1c
.RI "\fBcint\fP \fBwriteToLogDebug\fP (const std::string_view logmessg, const std::string_view logmodule=\fBAELOG_DEFAULT_MODULE_NAME\fP)"
.br
.RI "Request a debug log entry to be written to the opened log file\&. "
.ti -1c
.RI "\fBcint\fP \fBwriteStatus\fP (void)"
.br
.RI "Writes the current status of the file logger instance (file name, and entries written)\&. "
.ti -1c
.RI "std::string \fBgetLogName\fP (void) const noexcept"
.br
.RI "Get the name of the log file\&. "
.ti -1c
.RI "std::string \fBgetLogPath\fP (void) const"
.br
.RI "Get the path of the log file\&. "
.ti -1c
.RI "std::string \fBgetLogAbsolutePath\fP (void) const"
.br
.RI "Get the full/absolute path of the log file\&. "
.ti -1c
.RI "\fBullint\fP \fBgetEntryCount\fP (void) const noexcept"
.br
.RI "Get the amount of log entries done to an opened log file\&. "
.ti -1c
.RI "bool \fBisOpen\fP (void) const noexcept"
.br
.RI "Checks if the current log file is open\&. "
.ti -1c
.RI "bool \fBisClosed\fP (void) const noexcept"
.br
.RI "Checks if a log file isn't open by this logger\&. "
.ti -1c
.RI "bool \fBisWriting\fP (void) const noexcept"
.br
.RI "Checks if the log-writing thread is running\&. "
.ti -1c
.RI "constexpr const \fBullint\fP \fBgetModuleID\fP (void) noexcept"
.br
.RI "Get the \fBid number of the instance\fP of the module\&. "
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static std::string \fBgenLogName\fP (const std::string_view logpref, const std::string_view logext='\&.log')"
.br
.RI "Creates a full name of the log file to open to be fed to the logger, in a format [prefix]_[current date][\&.extension] in a given directory\&. "
.ti -1c
.RI "static \fBullint\fP \fBgetModuleAmount\fP (void) noexcept"
.br
.RI "Get the \fBamount of the currently-alive instances of module T\fP\&. "
.ti -1c
.RI "static constexpr const std::string_view \fBgetModuleName\fP (void) noexcept"
.br
.RI "Get the \fBname of the module\fP\&. "
.ti -1c
.RI "static constexpr const \fBullint\fP \fBgetInstanceCounter\fP (void) noexcept"
.br
.RI "Get the current value of the \fBinstance counter\fP of the module\&. "
.in -1c
.SS "Protected Attributes"

.in +1c
.ti -1c
.RI "const \fBullint\fP \fBm_ullModuleId\fP"
.br
.RI "The ID number of the current module\&. "
.in -1c
.SS "Static Protected Attributes"

.in +1c
.ti -1c
.RI "static std::atomic< std::size_t > \fBm_ullModulesAlive\fP"
.br
.RI "Amount of the currently-alive (not destroyed) instances of the module\&. "
.ti -1c
.RI "static std::atomic< \fBullint\fP > \fBm_ullInstanceCounter\fP"
.br
.RI "Amount of the instances of the module ever created during the program's runtime\&. "
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "\fBAELogEntry\fP * \fBptrFromIndex\fP (\fBullint\fP num) noexcept"
.br
.RI "Checks for the given index number and returns the pointer in the queue of log entries\&. "
.ti -1c
.RI "void \fBwriteToLogDirectly\fP (const std::string_view logmessg, const \fBcint\fP logtype=\fBAELOG_TYPE_INFO\fP, const std::string_view logmodule=\fBAELOG_DEFAULT_MODULE_NAME\fP) noexcept"
.br
.RI "Writes to log directly, without thinking of the queue\&. "
.ti -1c
.RI "void \fBlogWriterThread\fP (void)"
.br
.RI "The function of the log writing thread to\&.\&.\&.read the entries, format them, write them, and clear them\&. "
.ti -1c
.RI "\fBREGISTER_MODULE\fP (\fBAELogger\fP)"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "\fBAEFileWriter\fP \fBm_fwLogger\fP"
.br
.RI "The file writer to actually write text to opened log file\&. "
.ti -1c
.RI "std::mutex \fBm_mtxAllocLock\fP"
.br
.RI "The mutex to lock when allocating new queue chunks\&. "
.ti -1c
.RI "std::vector< std::pair< \fBullint\fP, \fBAELogEntry\fP * > > \fBm_vAllocTable\fP"
.br
.RI "The vector to all the allocated queue pieces and total amounts allocated\&. "
.ti -1c
.RI "std::thread \fBm_trdWriter\fP"
.br
.RI "The thread object for the file writing thread to\&.\&.\&.write a log file separately\&. "
.ti -1c
.RI "std::atomic< \fBullint\fP > \fBm_ullFilledCount\fP"
.br
.RI "The amount of nodes filled (should we allocate more?) "
.ti -1c
.RI "std::atomic< \fBullint\fP > \fBm_ullNodeNumber\fP"
.br
.RI "The current node number the writeToLog is working with\&. "
.ti -1c
.RI "\fBullint\fP \fBm_ullQueueSize\fP"
.br
.RI "The amount of entry spaces/size of the queue\&. "
.ti -1c
.RI "\fBAELogEntry\fP * \fBm_lepQueue\fP"
.br
.RI "The pointer to the first item in the queue\&. "
.ti -1c
.RI "\fBAELogEntry\fP * \fBm_lepLastNode\fP"
.br
.RI "The pointer to the last item in the queue\&. "
.ti -1c
.RI "std::atomic< bool > \fBm_bRunTrd\fP"
.br
.RI "The flag to run the log-writing thread\&. "
.ti -1c
.RI "std::atomic< bool > \fBm_bQueueFilled\fP"
.br
.RI "The flag to show that the queue is populated\&. "
.in -1c
.SH "Detailed Description"
.PP 
The ArtyK's Engine's Logger module -- it manages the writing to the log files\&. 

A wrapper around AEFW functionality for writing, and \fBAELogEntry\fP for mass-formatting of entries to text
.PP
The log is written by requesting and filling the entry in the queue\&. The \fBAELogger\fP instance launches the separate thread that looks at the entries in the queue, retrieves and formats the data in them, and writes it to the file\&. Afterwards that entry in the queue is cleared\&.
.PP
Hungarian notation is lg\&. (m_lgMyLogger)
.PP
\fBTodo\fP
.RS 4
Implement copy constructors and copy assignment 
.PP
Add the ability to open the same log file/redirect the instance requests to the one that has it open first\&. 
.RE
.PP
\fBBug\fP
.RS 4
The queue can expand if it's too little\&. But\&.\&.\&.\&.I don't know how to shrink it\&. (working on it) 
.RE
.PP

.PP
Definition at line \fB64\fP of file \fBAELogger\&.hpp\fP\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "AELogger::AELogger (const std::string fname, const bool clearLog = \fRfalse\fP, const \fBullint\fP queuesize = \fR\fBAELOG_DEFAULT_QUEUE_SIZE\fP\fP)\fR [inline]\fP, \fR [explicit]\fP"

.PP
Class constructor -- it opens the file with the given name and starts logging to it\&. 
.PP
\fBNote\fP
.RS 4
Puts the file into the default log path location (AELOG_DEFAULT_LOG_PATH)
.RE
.PP
\fBParameters\fP
.RS 4
\fIfname\fP Name of the log file
.br
\fIclearLog\fP Flag to clear the log file if it exists instead of appending it
.br
\fIqueuesize\fP The size of the queue to create when creating \fBAELogger\fP instance
.RE
.PP

.PP
Definition at line \fB76\fP of file \fBAELogger\&.hpp\fP\&.
.SS "AELogger::AELogger (const std::string logpath, const std::string fname, const bool clearLog = \fRfalse\fP, const \fBullint\fP queuesize = \fR\fBAELOG_DEFAULT_QUEUE_SIZE\fP\fP)\fR [explicit]\fP"

.PP
Class constructor -- it opens the file with the given path and name and start logging to it\&. 
.PP
\fBParameters\fP
.RS 4
\fIlogpath\fP The path of the log file to open it in
.br
\fIfname\fP Name of the log file
.br
\fIclearLog\fP Flag to clear the log file if it exists instead of appending it
.br
\fIqueuesize\fP The size of the queue to create when creating \fBAELogger\fP instance
.RE
.PP

.PP
Definition at line \fB13\fP of file \fBAELogger\&.cpp\fP\&.
.PP
References \fBAELOG_DEFAULT_ALLOC_VECTOR_RESERVE\fP, \fBAELOG_TYPE_OK\fP, \fB__AEModuleBase< AELogger >::getModuleName()\fP, \fBAEFileWriter::isOpen()\fP, \fBm_fwLogger\fP, \fBm_lepQueue\fP, \fBm_vAllocTable\fP, \fBstartWriter()\fP, and \fBwriteToLog()\fP\&.
.SS "AELogger::AELogger (void)\fR [inline]\fP, \fR [explicit]\fP"

.PP
Default class constructor -- doesn't open the file, but setups the instance for one\&. 
.PP
Definition at line \fB91\fP of file \fBAELogger\&.hpp\fP\&.
.PP
References \fBAELOG_DEFAULT_ALLOC_VECTOR_RESERVE\fP, \fBm_lepQueue\fP, \fBm_ullQueueSize\fP, and \fBm_vAllocTable\fP\&.
.SS "AELogger::~AELogger (void)"

.PP
Class destructor\&. 
.PP
Definition at line \fB33\fP of file \fBAELogger\&.cpp\fP\&.
.PP
References \fBAELOG_TYPE_INFO\fP, \fBcloseLog()\fP, \fB__AEModuleBase< AELogger >::getModuleName()\fP, \fBm_vAllocTable\fP, and \fBwriteToLog()\fP\&.
.SS "AELogger::AELogger (const \fBAELogger\fP &)\fR [delete]\fP"

.PP
Deleted (for now) copy-constructor\&. 
.SH "Member Function Documentation"
.PP 
.SS "\fBAELogger\fP & AELogger::operator= (const \fBAELogger\fP &)\fR [delete]\fP"

.PP
Deleted (for now) copy-assignment\&. 
.SS "\fBcint\fP AELogger::startWriter (void)"

.PP
Starts the log-writing thread\&. 
.PP
\fBReturns\fP
.RS 4
AELOG_ERR_NOERROR on success; otherwise AELOG_ERR_THREAD_ALREADY_RUNNING if thread already was running, AELOG_ERR_UNABLE_START_THREAD if error happened (+ std::runtime_error() exception)
.RE
.PP

.PP
Definition at line \fB44\fP of file \fBAELogger\&.cpp\fP\&.
.PP
References \fBAELOG_ERR_NOERROR\fP, \fBAELOG_ERR_THREAD_ALREADY_RUNNING\fP, \fBAELOG_ERR_UNABLE_START_THREAD\fP, \fBAELOG_TYPE_FATAL_ERROR\fP, \fBAELOG_TYPE_INFO\fP, \fBAELOG_TYPE_WARN\fP, \fBcloseLog()\fP, \fB__AEModuleBase< AELogger >::getModuleName()\fP, \fBlogWriterThread()\fP, \fBm_bRunTrd\fP, \fBm_trdWriter\fP, and \fBwriteToLog()\fP\&.
.SS "\fBcint\fP AELogger::stopWriter (void)"

.PP
Stops the log-writing thread (after flushing the log queue)\&. 
.PP
\fBReturns\fP
.RS 4
AELOG_ERR_NOERROR on success; otherwise AELOG_ERR_THREAD_ALREADY_STOPPED if thread already was stopped
.RE
.PP

.PP
Definition at line \fB63\fP of file \fBAELogger\&.cpp\fP\&.
.PP
References \fBAELOG_ERR_NOERROR\fP, \fBAELOG_ERR_THREAD_ALREADY_STOPPED\fP, \fBAELOG_TYPE_ERROR\fP, \fBAELOG_TYPE_INFO\fP, \fBAELOG_TYPE_OK\fP, \fB__AEModuleBase< AELogger >::getModuleName()\fP, \fBm_bRunTrd\fP, \fBm_trdWriter\fP, \fBwriteToLog()\fP, and \fBwriteToLogDirectly()\fP\&.
.SS "\fBcint\fP AELogger::openLog (const std::string fpath, const std::string fname, const bool clearLog = \fRfalse\fP)\fR [inline]\fP"

.PP
Open the file to start logging\&. 
.PP
\fBParameters\fP
.RS 4
\fIfpath\fP Path to put the log file in
.br
\fIfname\fP Name of the log file
.br
\fIclearLog\fP Flag to clear the log file if it exists instead of appending it
.RE
.PP
\fBReturns\fP
.RS 4
AELOG_ERR_NOERROR on success; otherwise return values of \fBAEFileWriter::openFile()\fP or \fBAELogger::startWriter()\fP
.RE
.PP

.PP
Definition at line \fB136\fP of file \fBAELogger\&.hpp\fP\&.
.PP
References \fBAEFW_ERR_NOERROR\fP, \fBAEFW_FLAG_APPEND\fP, \fBAELOG_TYPE_SUCCESS\fP, \fB__AEModuleBase< AELogger >::getModuleName()\fP, \fBm_fwLogger\fP, \fBAEFileWriter::openFile()\fP, \fBstartWriter()\fP, and \fBwriteToLog()\fP\&.
.SS "\fBcint\fP AELogger::openLog (const std::string fname, const bool clearLog = \fRfalse\fP)\fR [inline]\fP"

.PP
Open the file to start logging\&. 
.PP
\fBNote\fP
.RS 4
Puts the file into the default log path location (AELOG_DEFAULT_LOG_PATH)
.RE
.PP
\fBParameters\fP
.RS 4
\fIfname\fP Name of the log file
.br
\fIclearLog\fP Flag to clear the log file if it exists instead of appending it
.RE
.PP
\fBReturns\fP
.RS 4
AELOG_ERR_NOERROR on success; otherwise return values of \fBAEFileWriter::openFile()\fP or \fBAELogger::startWriter()\fP
.RE
.PP

.PP
Definition at line \fB152\fP of file \fBAELogger\&.hpp\fP\&.
.PP
References \fBAELOG_DEFAULT_LOG_PATH\fP, and \fBopenLog()\fP\&.
.SS "\fBcint\fP AELogger::openLog (const std::string_view fpath, const std::string_view fname, const bool clearLog = \fRfalse\fP)\fR [inline]\fP"

.PP
Open the file to start logging\&. 
.PP
\fBParameters\fP
.RS 4
\fIfpath\fP Path of directory to put the log file in (include trailing '/' character)
.br
\fIfname\fP Name of the log file
.br
\fIclearLog\fP Flag to clear the log file if it exists instead of appending it
.RE
.PP
\fBReturns\fP
.RS 4
AELOG_ERR_NOERROR on success; otherwise return values of \fBAEFileWriter::openFile()\fP or \fBAELogger::startWriter()\fP
.RE
.PP

.PP
Definition at line \fB163\fP of file \fBAELogger\&.hpp\fP\&.
.PP
References \fBopenLog()\fP\&.
.SS "\fBcint\fP AELogger::closeLog (void)\fR [inline]\fP"

.PP
Close the file, if it was opened\&. That's it\&.
.PP
\fBReturns\fP
.RS 4
AELOG_ERR_NOERROR on success; otherwise AEFW_ERR_FILE_NOT_OPEN if file isn't open, 
.RE
.PP

.PP
Definition at line \fB171\fP of file \fBAELogger\&.hpp\fP\&.
.PP
References \fBAEFW_ERR_FILE_NOT_OPEN\fP, \fBAELOG_ERR_NOERROR\fP, \fBAELOG_TYPE_INFO\fP, \fBAELOG_TYPE_SUCCESS\fP, \fBAEFileWriter::closeFile()\fP, \fBAEFileWriter::getFullFileName()\fP, \fBisClosed()\fP, \fBm_fwLogger\fP, \fBstopWriter()\fP, \fBwriteToLog()\fP, and \fBwriteToLogDirectly()\fP\&.
.SS "\fBcint\fP AELogger::writeToLog (const std::string_view logmessg, const \fBcint\fP logtype = \fR\fBAELOG_TYPE_INFO\fP\fP, const std::string_view logmodule = \fR\fBAELOG_DEFAULT_MODULE_NAME\fP\fP, const std::time_t logTime = \fR0\fP)"

.PP
Request a log entry to be written to the opened log file\&. 
.PP
\fBNote\fP
.RS 4
See AELOG_TYPE_* flags 
.PP
The module name should contain only alphanumeric characters or underscores (no spaces), otherwise it fails
.RE
.PP
\fBParameters\fP
.RS 4
\fIlogmessg\fP The message of the requested log entry
.br
\fIlogtype\fP The type of the log entry
.br
\fIlogmodule\fP The name of the module that invoked this request
.br
\fIlogTime\fP The custom time of the log entry to insert, \fIif you really need that\fP!
.RE
.PP
\fBReturns\fP
.RS 4
AELOG_ERR_NOERROR on success; otherwise AEFW_ERR_FILE_NOT_OPEN if log file isn't open, AELOG_ERR_INVALID_ENTRY_DATA if passed data isn't of proper format
.RE
.PP

.PP
\fBTodo\fP
.RS 4
Implement decrease in log queue size\&.\&.\&.somehow 
.RE
.PP

.PP
Definition at line \fB78\fP of file \fBAELogger\&.cpp\fP\&.
.PP
References \fBAEFW_ERR_FILE_NOT_OPEN\fP, \fBAELE_MESSAGE_SIZE\fP, \fBAELE_MODULENAME_SIZE\fP, \fBAELE_STATUS_INVALID\fP, \fBAELE_STATUS_READY\fP, \fBAELE_STATUS_SETTING\fP, \fBAELOG_ERR_INVALID_ENTRY_DATA\fP, \fBAELOG_ERR_NOERROR\fP, \fBAELOG_TYPE_DEBUG\fP, \fBAELOG_TYPE_FATAL_ERROR\fP, \fB__AEModuleBase< AELogger >::getModuleName()\fP, \fBace::utils::isAlNumUs()\fP, \fBisClosed()\fP, \fBace::utils::isInRange()\fP, \fBAELogEntry::m_cLogType\fP, \fBAELogEntry::m_cStatus\fP, \fBm_lepLastNode\fP, \fBm_lepQueue\fP, \fBm_mtxAllocLock\fP, \fBAELogEntry::m_pNextNode\fP, \fBAELogEntry::m_sLogMessage\fP, \fBAELogEntry::m_sModuleName\fP, \fBAELogEntry::m_tmLogTime\fP, \fBm_ullFilledCount\fP, \fBm_ullNodeNumber\fP, \fBm_ullQueueSize\fP, \fBm_vAllocTable\fP, \fBAELogEntry::makeQueue()\fP, \fBptrFromIndex()\fP, and \fBwriteToLogDebug()\fP\&.
.SS "\fBcint\fP AELogger::writeToLog (const \fBAELogEntry\fP & entry, const bool useCurrentTime = \fRtrue\fP)\fR [inline]\fP"

.PP
Request a log entry to be written to the opened log file, using data from a log entry type\&. 
.PP
\fBParameters\fP
.RS 4
\fIentry\fP The log entry to write to the file
.br
\fIuseCurrentTime\fP Flag to use current time for the log entry, or use timestamp in the provided entry
.RE
.PP
\fBReturns\fP
.RS 4
AELOG_ERR_NOERROR on success; otherwise AEFW_ERR_FILE_NOT_OPEN if log file isn't open
.RE
.PP

.PP
Definition at line \fB203\fP of file \fBAELogger\&.hpp\fP\&.
.PP
References \fBAELogEntry::m_cLogType\fP, \fBAELogEntry::m_sLogMessage\fP, \fBAELogEntry::m_sModuleName\fP, \fBAELogEntry::m_tmLogTime\fP, and \fBwriteToLog()\fP\&.
.SS "\fBcint\fP AELogger::writeToLogDebug (const std::string_view logmessg, const std::string_view logmodule = \fR\fBAELOG_DEFAULT_MODULE_NAME\fP\fP)\fR [inline]\fP"

.PP
Request a debug log entry to be written to the opened log file\&. 
.PP
\fBNote\fP
.RS 4
See AELOG_TYPE_* flags 
.PP
If ENGINE_DEBUG flag is not set, doesn't do anything 
.RE
.PP
\fBSee also\fP
.RS 4
\fBAELogger::writeToLog()\fP
.RE
.PP
\fBParameters\fP
.RS 4
\fIlogmessg\fP The message of the requested log entry
.br
\fIlogmodule\fP The name of the module that invoked this request
.RE
.PP
\fBReturns\fP
.RS 4
AELOG_ERR_NOERROR on success; otherwise AEFW_ERR_FILE_NOT_OPEN if log file isn't open, AELOG_ERR_INVALID_ENTRY_DATA if passed data isn't of proper format
.RE
.PP

.PP
Definition at line \fB216\fP of file \fBAELogger\&.hpp\fP\&.
.PP
References \fBAELOG_TYPE_DEBUG\fP, and \fBwriteToLog()\fP\&.
.SS "\fBcint\fP AELogger::writeStatus (void)\fR [inline]\fP"

.PP
Writes the current status of the file logger instance (file name, and entries written)\&. 
.PP
\fBReturns\fP
.RS 4
AELOG_ERR_NOERROR on success; otherwise AEFW_ERR_FILE_NOT_OPEN if log file isn't open, AELOG_ERR_INVALID_ENTRY_DATA if passed data isn't of proper format
.RE
.PP

.PP
Definition at line \fB227\fP of file \fBAELogger\&.hpp\fP\&.
.PP
References \fBAELOG_TYPE_INFO\fP, \fBAEFileWriter::getFullFileName()\fP, \fBm_fwLogger\fP, and \fBwriteToLog()\fP\&.
.SS "std::string AELogger::getLogName (void) const\fR [inline]\fP, \fR [noexcept]\fP"

.PP
Get the name of the log file\&. 
.PP
\fBReturns\fP
.RS 4
std::string of the name of opened log file; otherwise values from \fBAEFileWriter::getFullFileName()\fP
.RE
.PP

.PP
Definition at line \fB237\fP of file \fBAELogger\&.hpp\fP\&.
.PP
References \fBAEFileWriter::getFullFileName()\fP, and \fBm_fwLogger\fP\&.
.SS "std::string AELogger::getLogPath (void) const\fR [inline]\fP"

.PP
Get the path of the log file\&. 
.PP
\fBReturns\fP
.RS 4

AELogger/src/AELogger.cpp

/*! @file AELogger/src/AELogger.cpp
 *  This file contains the definitions of the engine's logger's functions.
 *  @see AELogger
 *  @see AELogger.hpp
 *
 *  Should not cause everything to break :)
 */

#include "../include/AELogger.hpp"

//constructor
AELogger::AELogger(const std::string logpath, const std::string fname, const bool clearLog, const ullint queuesize) :
	m_fwLogger(logpath + fname, !clearLog * AEFW_FLAG_APPEND /* Funny magic with bool-int conversion */),
	m_ullFilledCount(0), m_ullNodeNumber(0), m_ullQueueSize(queuesize),
	m_lepQueue(AELogEntry::makeQueue(queuesize, true, nullptr)), m_lepLastNode(m_lepQueue + queuesize - 1),
	m_bRunTrd(false), m_bQueueFilled(false) {


	//add the allocated queue pointed to the list
	//so we can free them later without memory leaks
	this->m_vAllocTable.reserve(AELOG_DEFAULT_ALLOC_VECTOR_RESERVE);
	this->m_vAllocTable.emplace_back(queuesize, this->m_lepQueue);

	if (this->m_fwLogger.isOpen()) {
		this->writeToLog("Created the AELogger instance and opened the log session in the file: \"" + std::filesystem::absolute(fname).generic_string() + '\"', AELOG_TYPE_OK, this->getModuleName());
		this->startWriter();
	}

}

//destructor
AELogger::~AELogger(void) {
	this->writeToLog("Destroying the AELogger instance and closing", AELOG_TYPE_INFO, this->getModuleName());
	this->closeLog();

	for (std::size_t i = 0; i < this->m_vAllocTable.size(); i++) {
		delete[] this->m_vAllocTable[i].second;
	}
	this->m_vAllocTable.clear();
}


cint AELogger::startWriter(void) {
	this->writeToLog("Attempting to start the log-writing thread", AELOG_TYPE_INFO, this->getModuleName());
	if (this->m_trdWriter.joinable()) {
		this->writeToLog("The log-writing thread is already running!", AELOG_TYPE_WARN, this->getModuleName());
		return AELOG_ERR_THREAD_ALREADY_RUNNING; //we already are writing, dummy;
	}

	this->m_bRunTrd = true;
	this->m_trdWriter = std::thread(&AELogger::logWriterThread, this);
	if (!this->m_trdWriter.joinable()) {
		this->writeToLog("Could not start AELogger thread!! (exception)", AELOG_TYPE_FATAL_ERROR, this->getModuleName());
		this->closeLog();
		throw std::runtime_error("Could not start AELogger thread!");
		return AELOG_ERR_UNABLE_START_THREAD;
	}
	return AELOG_ERR_NOERROR;
}


cint AELogger::stopWriter(void) {

	this->writeToLog("Attempting to stop the log-writing thread...", AELOG_TYPE_INFO, this->getModuleName());
	this->m_bRunTrd = false;
	if (this->m_trdWriter.joinable()) {
		this->m_trdWriter.join();
		
		this->writeToLogDirectly("Successfully exited the writer thread.", AELOG_TYPE_OK, this->getModuleName()); // the log is closed, so is the queue
		return AELOG_ERR_NOERROR;
	}
	this->writeToLogDirectly("Cannot close the already-closed thread!", AELOG_TYPE_ERROR, this->getModuleName());
	return AELOG_ERR_THREAD_ALREADY_STOPPED;
}

// request a log entry and write to it
cint AELogger::writeToLog(const std::string_view logmessg, const cint logtype, const std::string_view logmodule, const std::time_t logTime) {
	if (this->isClosed()) {
		return AEFW_ERR_FILE_NOT_OPEN; // file's closed/closing!
	}

	// The only reason this function isn't shortcutted to a writeToLog with an entry
	// is simply because it is doing the same thing anyway
	// copying string data to a log entry
	// and we'll need to do the same thing if we want to use the 2nd version of writeToLog
	
	/// @todo Implement decrease in log queue size...somehow

	// check for invalid arguments
	if (logmessg.empty() || logmodule.empty() || !ace::utils::isInRange(AELOG_TYPE_DEBUG, AELOG_TYPE_FATAL_ERROR, logtype) || //empty/invalid stuff
		!ace::utils::isAlNumUs(logmodule)) { //log module stuff
		return AELOG_ERR_INVALID_ENTRY_DATA;
	}

	this->m_ullFilledCount++;
	

	if (this->m_ullFilledCount > this->m_ullQueueSize) {
		const std::lock_guard<std::mutex> lock(this->m_mtxAllocLock);
		const ullint qsize = this->m_ullQueueSize / 2;

		//make new queue
		AELogEntry* newQueue = AELogEntry::makeQueue(qsize, true, this->m_lepQueue);
		//add the new queue to the vector
		this->m_ullQueueSize += qsize;
		this->m_lepLastNode->m_pNextNode = newQueue;
		this->m_lepLastNode = newQueue + qsize - 1;
		this->m_vAllocTable.emplace_back(this->m_ullQueueSize, newQueue);
		this->writeToLogDebug("The queue was too small, resized it to " + std::to_string(this->m_ullQueueSize) + " entries", this->getModuleName());
	}

	
	//implementation: the allocation vector!
	//instead of it having only the pointers to the allocated memory chunks of the log entries
	//it would have the std::pair instead: pointer to chunk and the total amount of entries
	//ex: {0xDEAD, 1024}, {0xBEEF, 1096} -> first pointer holds 1024 total entries, second holds 1096-1024=62 entries
	//and instead of changing an atomic pointer, we'll increment a number, "current node number"
	//and then check if the number(modulo of it with the m_ullQueueSize) is less than the 1st allocation in the vector, then 2nd, etc..untill we find a match
	//then, to get the pointer to current node -> pointer of that allocation + number


	//increment node number and get the pointer
	AELogEntry* ptr = this->ptrFromIndex(m_ullNodeNumber++);
	while (ptr->m_cStatus != AELE_STATUS_INVALID) {
		ptr = this->ptrFromIndex(m_ullNodeNumber++); //if current node is filled -> continue looking for unpopulated one
	}

	
	//populating it now!
	ptr->m_cStatus = AELE_STATUS_SETTING; //alright boys, we're setting this one up

	ptr->m_tmLogTime = (logTime == 0) ? std::time(nullptr) : logTime;
	memcpy(ptr->m_sLogMessage, logmessg.data(), (logmessg.size() > AELE_MESSAGE_SIZE) ? AELE_MESSAGE_SIZE : logmessg.size());
	memcpy(ptr->m_sModuleName, logmodule.data(), (logmodule.size() > AELE_MODULENAME_SIZE) ? AELE_MODULENAME_SIZE : logmodule.size());
	ptr->m_cLogType = logtype;
	ptr->m_cStatus = AELE_STATUS_READY;
	return AELOG_ERR_NOERROR;
}



void AELogger::logWriterThread(void) {
	//bit of announcements
	this->writeToLog("Successfully launched the log-writing thread!", AELOG_TYPE_SUCCESS, this->getModuleName());


	//pointers to traverse the queue
	AELogEntry* ePtr = this->m_lepQueue;
	AELogEntry* oldptr = ePtr;

	

	//the final message to output
	char str[AELE_FORMAT_MAX_SIZE]{};

	// The node order number for the writing thread.
	//ullint m_ullWriterOrderNum = 0;


	//and not stop untill it's done
	//untill we written everything *and* we stopped the thread
	

	while (this->m_bRunTrd.load(std::memory_order::relaxed) || this->m_ullFilledCount.load(std::memory_order::relaxed)) {
		while (this->m_ullFilledCount.load(std::memory_order::relaxed)) {
			
			//got it!. Now wait untill it's ready
			while (ePtr->m_cStatus != AELE_STATUS_READY) {
				ace::utils::sleepUS(100);
			}
			//the entry is minee!
			ePtr->m_cStatus = AELE_STATUS_READING;
			
			//formatting
			
			AELogEntry::formatEntry(str, *ePtr);
			
			//std::cout << str;
			
			//writing
			this->m_fwLogger.writeData_ptr(str, std::strlen(str), 1, false);
			this->m_fwLogger.flushFile();
			
			//gotta go to the next ptr
			ePtr = ePtr->m_pNextNode;

			//cleanup
			AELogEntry::clearEntry(*oldptr);
			oldptr = ePtr;
			this->m_ullFilledCount--;
			//m_ullWriterOrderNum++;
			
		}

		//since we're done writing the entries, sleep and check if some appear again
		ace::utils::sleepUS(1000);
	}
	this->writeToLogDirectly("Reached the end of the log-writing thread function", AELOG_TYPE_OK, this->getModuleName());
}


AELogEntry* AELogger::ptrFromIndex(ullint num) noexcept {
	num %= m_ullQueueSize;
	ullint prevSize = 0;
	for (std::size_t i = 0; i < this->m_vAllocTable.size(); i++) {
		if (this->m_vAllocTable[i].first > num) {
			return this->m_vAllocTable[i].second + (num - prevSize);
		}
		prevSize = this->m_vAllocTable[i].first;
	}
	return m_lepQueue;
}



